### PR TITLE
implement dynamic batching

### LINKED ISSUE #4: Add batching

### PR DESCRIPTION
**Pull request by aniketmaurya**
<details>
  <summary><b>Before submitting</b></summary>

- [ ] Was this discussed/agreed via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

</details>

## What does this PR do?

Fixes #4 .

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in GitHub issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by lantiga** [comment]
Hey @aniketmaurya, can you take a look at this branch? https://github.com/Lightning-AI/litserve/compare/main...luca-batching

File: src/litserve/api.py [hunk 1]
```diff
@@ -3,6 +3,21 @@
 from abc import ABC, abstractmethod
 
 
+def no_batch_unbatch_message(obj):
+    return f"""
+        You set `max_batch_size > 1` but didn't implement batch() and unbatch().
```
Comment (lantiga) [inline src/litserve/api.py:8]: We are providing a default batch and unbatch implementation for tensors and ndarrays, so we could improve the messaging to be more specific.

```
You set `max_batch_size > 1`, but the default implementation for batch() and unbatch() only supports PyTorch tensors or NumPy ndarrays, while we found {TODO ADD TYPE HERE}.
Please implement these two methods in {obj.__class__.__name__}.
```

File: src/litserve/server.py [hunk 2]
```diff
@@ -19,15 +20,59 @@
...
+
+
+def collate_requests(lit_api, request_queue: Queue, request_buffer, max_batch_size, batch_timeout):
+    uids = []
+    entered_at = time.time()
+    while (batch_timeout - (time.time() - entered_at) > 0) and len(uids) <= max_batch_size:
+        try:
+            uid = request_queue.get_nowait()
+            uids.append(uid)
+        except (Empty, ValueError):
```
Comment (lantiga) [inline src/litserve/server.py:42]: Should we add a `time.sleep(0.001)` to avoid looping too aggressively?

File: src/litserve/server.py [hunk 3]
```diff
@@ -19,15 +20,59 @@
...
+
+
+def run_single_loop(lit_api, request_queue: Queue, request_buffer):
     while True:
-        # NOTE: to implement batching here: keep getting items from the queue,
-        #       fill a batch, predict, send outputs to the respective pipes
-        #       In the future we will expose this through the API.
         try:
-            uid = request_queue.get(timeout=1.0)
+            uid = request_queue.get(timeout=0.01)
```
Comment (lantiga) [inline src/litserve/server.py:75]: Is there a reason why we set the timeout to be so low? It will unblock anyway when something shows up on the queue.

**Review by lantiga — APPROVED** [review]
Couple of comments, looks great otherwise

File: src/litserve/server.py (hunk 3 above)
Comment (aniketmaurya) [reply src/litserve/server.py]: yeah make sense. reverted to `1.0`

File: src/litserve/api.py (hunk 1 above)
Comment (aniketmaurya) [reply src/litserve/api.py]: updated to this ^

File: src/litserve/server.py (hunk 2 above)
Comment (aniketmaurya) [reply src/litserve/server.py:42]: how about `uid = request_queue.get(timeout=0.001)z

**Comment by aniketmaurya** [comment]
@lantiga updated the PR based on your feedback.

File: src/litserve/server.py (hunk 2 above)
Comment (lantiga) [reply src/litserve/server.py:42]: Ok let's do that

**Pull request merged**